astro/utils/parallel.go

```diff
@@ -17,21 +17,41 @@
 package utils
 
 import (
+	"fmt"
+	"os"
+	"os/signal"
 	"sync"
+	"syscall"
 )
 
 // Parallel runs at most maxConcurrent functions in parallel.
 func Parallel(maxConcurrent int, fns ...func()) {
 	wg := sync.WaitGroup{}
 	guard := make(chan struct{}, maxConcurrent)
+
+	isShutDown := false
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, syscall.SIGTERM, os.Interrupt)
+	go func() {
+		signal := <-signalCh
+		fmt.Printf("\nReceived signal: %s, gracefully shutting down...\n", signal)
+		isShutDown = true
+	}()
+
 	for _, fn := range fns {
-		guard <- struct{}{}
-		wg.Add(1)
-		go func(fn func()) {
-			defer wg.Done()
-			fn()
-			<-guard
-		}(fn)
+		// https://stackoverflow.com/a/25306241
+		guard <- struct{}{} // would block if guard channel is already filled
+		if isShutDown {
+			fmt.Println("Shutting down. No new terraform processes will be launched.")
```

Comment:
This function is a utility function and has nothing to do with Terraform.

Do you think it's an acceptable design to make `Parallel` stop executing fns on SIGTERM? It kinda feels like Parallel should be more general than that.

What happened to the idea of return a context allowing the caller can cancel it?

Reply:
Now it receives a context with cancel